### PR TITLE
Expose runtime baking functionality in LightmapGI

### LINKED ISSUE #8656: Expose method `LightmapGI.bake()` in editor and exported projects
### Describe the project you are working on

I'm working on a procedural and UGC *(User Generated Content)* game where users can create their own maps from premade blocks and custom models.

### Describe the problem or limitation you are having in your project

I need to have lightmaps for performance and look reasons as I'm targeting both low and high end PCs.  
As it is a game where users can create their own maps from premade blocks (or their own) in a map editor ingame this mean I cannot use the editor for that.  

Since GPUs become faster and faster everyday and that LightmapGI uses GPU for baking, this is perfect for this proposal!  

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Exposing the method help for procedural and UGC games:
- The user wouldn't need to download a whole editor and a project just to bake lights of their maps, it would be seamless ingame.
- For procedural scenes that require realistic lighting or lighting which cannot be provided by SDFGI or VoxelGI alone, having lightmaps unlock that.

Having the bake method exposed also help for the editor side:
- You could have a script which bake multiple scenes at a time, saving time and letting you do other things (such as sleeping if you're baking a dozen of maps at high quality).
- You could automate and test lightmap changes (such as bounce, bounce feedback, ...) and which output a screenshot, saving a lot of time and effort.
- You could have a tool which apply some stuff before baking and restore them later (eg: occluders that should be visible in the bake but hidden after that)

### Code A:
Based on how LightmapGIEditorPlugin does it:
```gdscript
func on_bake_step(step_progress: float, step_description: String) -> bool:
    return Input.is_key_pressed(KEY_ESCAPE) # Cancel if the user pressed the escape key

# LightmapGI.bake(bake_step_callback: Callable, from_node: Node = null)
var err: BakeError = lm.bake(on_bake_step)
```

### Code B:
More understandable and traditional for developers:
```gdscript
func start_baking():
    # LightmapGI.bake_start(from_node: Node = null)
    var err: BakeError = lm.bake_start() # If there is an error, the baking will stop immediately

func _process(delta):
    if lm.is_baking:
        # LightmapGI.bake_status() -> {step_progress=0.0, step_description="", error=BakeError}
        var status: Dictionary = lm.bake_status()
        if Input.is_key_pressed(KEY_ESCAPE): # Cancel if the user pressed the escape key
            lm.bake_abort()
```
*Code B is my preferred choice and make more sense for developers, especially if you check the next note ↓*  

***Note:** as of right now, I'm not sure how Godot does to call `process` while the lightmap is baking as it should be blocking by looking at the code (but throwing a tool script it does call process...), so the method where you call it is blocking which could cause problems if badly implemented on the game code (such as an infinite loop).*  
***Note:** the `bake` or `bake_start` require a `LightmapGIData` to be set in the node or else it should return an error. Perhaps passing one in the method parameter could be done.*

##
As of right now, it's not possible to cancel lightmapping (even thought it has a dedicated error) but I think **canceling it should be implemented**, mainly for those reasons:
- If you do a batch bake in the editor, you need to be able to cancel or else you may loose the progress of previous scenes bake.
- For a game, if the user want to quit the game or abort the baking process they should be able to do so and not resort to kill the game process.
##
As I've read on this issue https://github.com/godotengine/godot/issues/59217 (which is only for exposing it to editor) this would increase the binary size on exported projects:  
To solve that, a feature would be added in the `Build Configuration Profile` window with the name `Lightmapping`. It would be disabled by default and would require the game maker to enable it manually.  
If disabled and the `bake` method is called, an error will pop out in the console indicating what the user should do and `bake` would return `BakeError.BAKE_ERROR_NO_LIGHTMAPPER`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

LightmapGI is core.

### LINKED ISSUE #59217: Vulkan: LightmapGI's `bake()` function is not exposed to scripting (unlike BakedLightmap in `3.x`)
### Godot version

4.0.alpha (fb2802578)

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 495.46)

### Issue description

LightmapGI's `bake()` function is not exposed to scripting (unlike BakedLightmap in `3.x`).

Exposing `bake()` is useful for editor plugins that need to perform lightmap baking in an automatic manner, without relying on the BakedLightmap node currently being selected by the user.

Note that `bake()` will still not work in exported projects, as lightmapping functionality is only available in the editor to decrease binary size. `bake()` will only work in projects run from the editor, as well as the editor itself.

I've tried to do this here, but I can't get environment lighting to show up when baking lightmaps by calling `bake()` in the editor on scene load: https://github.com/Calinou/godot/tree/lightmapgi-expose-bake-method

### Steps to reproduce

- Look at the LightmapGI class documentation.
- Notice there is no `bake()` method exposed to scripting.

### Minimal reproduction project

Testing project for the above linked branch: [test_lightmapgi_bake_method.zip](https://github.com/godotengine/godot/files/8275427/test_lightmapgi_bake_method.zip)

### PR DESCRIPTION
**Pull request by RisingThumb**
This will resolve https://github.com/godotengine/godot/issues/59217
This partially resolves https://github.com/godotengine/godot-proposals/issues/8656 . It exposes `bake()` for scripting, however there isn't any cancelling or aborting yet. ~~Additionally it does not currently work for exported projects.~~ It also works for exported projects

Implementation notes:

- As .exr files cannot be imported at runtime, I save as a .res. This is also because when the lightmapper is available in exported projects, we can use .res files, while we can't save .exr files
- The radiance changes setting to Radiance 128, and then resetting back to what it was, is because the lightmaps were not generating correctly at runtime using the Scene/Custom Sky environment options. Setting it to 128 fixes this. I am not sure what is causing this, possibly something related to `environment_bake_panorama` or `sky_bake_panorama`. There is a related Size2i parameter for this radiance size, but setting that to 256 while using a radiance size of 256 doesn't fix this issue.
- Is there a better way of converting from an image to a texturelayered?

### CODE REVIEW

**Comment by jcostello** [comment]
In which cases you will bake lightmap in runtime?

**Comment by passivestar** [comment]
> In which cases you will bake lightmap in runtime?

In custom level editors built with godot

Also when [reimporting levels in godot while the game is running](https://github.com/godotengine/godot-proposals/issues/9519). Here's me taking advantage of the fact that VoxelGI can be baked at runtime: https://twitter.com/passivestar_/status/1780433192211612039

**Comment by RisingThumb** [comment]
> In which cases you will bake lightmap in runtime?

This opens up the opportunity for having a script that goes through all your maps/scenes and baking these lightmaps, rather than doing it in-editor, since the editor process is manual currently. So if you want a lot of high quality lightmaps for a lot of scenes, you can go away and do something else in the meantime. Additionally, it also opens it up for custom map makers in a similar way to Quake, where the lightmaps need to be baked(however this would require additional work for it to work in exports as the lightmapper).

> Note that bake() will still not work in exported projects, as lightmapping and automatic UV unwrapping (xatlas) functionality is only available in the editor to decrease binary size. bake() will only work in projects run from the editor, as well as the editor itself.

That said this is currently for runtime with the editor, not currently in exported builds(quoted from #59217 ), so it doesn't get us all the way there yet, but at least to the same state Godot 3.X was at which for projects like func_godot, is still very helpful

**Comment by nklbdev** [comment]
I'm joining the topicstarter.

I'm creating an engine for a game based on Godot, but the maps are outside the resource file system. They are located in the next folder with the executable file. They are created in TrenchBroom (*.map-files). I use my own parser for these files. Not [Quodot](https://github.com/QodotPlugin/qodot-plugin) and not [func_godot](https://github.com/func-godot). To build them, I call the game executable file via the command line:
```
> game.exe --headless -- build path/to/file.map
```
Here a packed scene is created and saved to another external `*.scn`-file. The scene stores the geometry, light, physics and location of the player, monsters, game objects as well as the conditions for switching to another map. The map file does not embed textures, materials and internal details of game objects.
Everything goes fine and after that the scene can be opened and played.

I really miss the ability to bake light in map building script.

**Comment by Calinou** [comment]
Please discuss the functionality itself in the [proposal](https://github.com/godotengine/godot-proposals/issues/8656), as comments on the pull request should only refer to the *implementation*.

**Comment by sourcelocation** [comment]
Hi, I tried compiling this fork and I was surprised to find that the exported project (at least the debug build, haven't checked release yet) actually *can* bake lightmaps into .res file format, however they don't look right. I assume it's due to slice counts. The first screenshot is a lightmap compiled from within the editor, and the second is from the game.

image
image

(Temporarily disabled denoiser)

What leads me to believe this is an issue with slices is the fact that if I change the vertical slice count from 2 to 1, I get a rather similar broken lightmap to the one I got while baking in the game. 
image

As there aren't any import options for .res files, I'm not sure what the issue could be here. I want to contribute, I just can't figure out this problem. Do you have an idea by any chance @Calinou?

**Comment by sourcelocation** [comment]
Update: I got it working in the exported project here https://github.com/RisingThumb/godot/pull/1. This is my first ever contribution to the Godot's codebase, so probably a lot of things are wrong there. I need some assistance.

**Comment by RisingThumb** [comment]
Thanks @sourcelocation , I've merged your changes into this PR which make it possible to bake at runtime for exported projects into this PR. Also means it addresses the original issue more completely now, as it works for exported projects if `module_lightmapper_rd_enabled` and `module_xatlas_unwrap_enabled` are enabled 👍

**Comment by sourcelocation** [comment]
When can we expect this PR to be reviewed?

**Comment by Calinou** [comment]
> When can we expect this PR to be reviewed?

4.3 is in feature freeze, so new features can only be merged in 4.4 at the earliest.

Also, there are merge conflicts that should be resolved. See [Pull request workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions :slightly_smiling_face:

File: doc/classes/LightmapGI.xml [hunk 1]
```diff
@@ -14,6 +14,17 @@
...
 		<link title="Using Lightmap global illumination">$DOCS_URL/tutorials/3d/global_illumination/using_lightmap_gi.html</link>
 	</tutorials>
+	<methods>
+		<method name="bake">
+			<return type="int" enum="LightmapGI.BakeError" />
+			<param index="0" name="from_node" type="Node" />
+			<param index="1" name="image_data_path" type="String" default="&quot;&quot;" />
+			<description>
+				Bakes lightmaps (requires meshes to have UV2 unwrapped) for [param from_node] and its children to [param image_data_path]. [param image_data_path] must end with an [code].exr[/code] or [code].lmbake[/code] file extension. If [param from_node] is [code]null[/code], lightmaps are baked from the [LightmapGI] node's parent. Baking lightmaps can take from a few seconds to several dozen minutes depending on the GPU speed and quality settings chosen.
+				[b]Note:[/b] [method bake] only works within the editor, and when running a project from the editor. [method bake] will do nothing when called in a project exported in either debug or release mode. This limitation is in place to reduce the binary size of exported projects.
```
Comment (Calinou) [inline doc/classes/LightmapGI.xml:24]:
```suggestion
				Bakes lightmaps (requires meshes to have UV2 unwrapped) for [param from_node] and its children to [param image_data_path]. [param image_data_path] must end with an [code].exr[/code] or [code].lmbake[/code] file extension. If [param from_node] is [code]null[/code], lightmaps are baked from the [LightmapGI] node's parent. Baking lightmaps can take from a few seconds to several dozen minutes depending on the GPU speed and quality settings chosen.
				[b]Note:[/b] [method bake] only works within the editor, and when running a project from the editor. [method bake] will do nothing when called in a project exported in either debug or release mode. This limitation is in place to reduce the binary size of exported projects. You can compile custom export templates with the [code]module_lightmapper_rd_enabled=yes module_xatlas_unwrap_enabled=yes[/code] SCons options to remove this limitation.
```

File: modules/lightmapper_rd/config.py [hunk 2]
```diff
@@ -1,6 +1,17 @@
...
+    from SCons.Script import BoolVariable, Variables, Help
+
+    envvars = Variables()
+    envvars.Add(
+        BoolVariable(
+            "lightmapper_rd",
+            "Enable Lightmapper functionality in export template builds (increases binary size)",
+            False,
+        )
+    )
```
Comment (Calinou) [inline modules/lightmapper_rd/config.py:15]: I'm confused why we need a separate `lightmapper_rd` option for this, considering we already have `module_lightmapper_rd_enabled` option that is `True` by default in editor builds and `False` by default in export template sbuilds.

File: modules/xatlas_unwrap/config.py [hunk 3]
```diff
@@ -1,6 +1,17 @@
...
+    from SCons.Script import BoolVariable, Variables, Help
+
+    envvars = Variables()
+    envvars.Add(
+        BoolVariable(
+            "xatlas_unwrap",
+            "Enable xatlas unwrapping functionality in export template builds (increases binary size)",
+            False,
+        )
+    )
```
Comment (Calinou) [inline modules/xatlas_unwrap/config.py:15]: Same as above.

Comment (sourcelocation) [reply modules/xatlas_unwrap/config.py:15]: Will work on it! Thanks.

**Comment by AThousandShips** [comment]
@RisingThumb please confirm you have abandoned this PR

**Comment by RisingThumb** [comment]
Hi, I'm happy for this PR to be superseded by https://github.com/godotengine/godot/pull/94965

**Pull request closed without merging**